include/oneapi/dpl/pstl/hetero/memory_ranges_impl_hetero.h

```diff
@@ -0,0 +1,244 @@
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_MEMORY_RANGES_IMPL_HETERO_H
+#define _ONEDPL_MEMORY_RANGES_IMPL_HETERO_H
+
+#if _ONEDPL_CPP20_RANGES_PRESENT
+
+#    include <ranges>
+#    include <utility>
+#    include <cassert>
+#    include <functional>
+#    include <type_traits>
+
+#    include "algorithm_ranges_impl_hetero.h" // for __pattern_walk_n
+#    include "dpcpp/utils_ranges_sycl.h"      // for oneapi::dpl::__internal::__ranges::views::all, etc.
+#    include "dpcpp/execution_sycl_defs.h"    // for __hetero_tag
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __internal
+{
+namespace __ranges
+{
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_uninitialized_default_construct
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R>
+std::ranges::borrowed_iterator_t<_R>
+__pattern_uninitialized_default_construct(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r)
+{
+    using _ValueType = typename std::ranges::range_value_t<_R>;
+
+    auto __last = std::ranges::begin(__r) + std::ranges::size(__r);
+
+    if constexpr (!std::is_trivially_default_constructible_v<_ValueType>)
+    {
+        oneapi::dpl::__internal::__ranges::__pattern_walk_n(
+            __tag, std::forward<_ExecutionPolicy>(__exec),
+            oneapi::dpl::__internal::__op_uninitialized_default_construct<std::decay_t<_ExecutionPolicy>>{},
+            oneapi::dpl::__ranges::views::all(std::forward<_R>(__r)));
+    }
+
+    return __last;
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_uninitialized_value_construct
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R>
+std::ranges::borrowed_iterator_t<_R>
+__pattern_uninitialized_value_construct(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r)
+{
+    using _ValueType = typename std::ranges::range_value_t<_R>;
+
+    auto __last = std::ranges::begin(__r) + std::ranges::size(__r);
+
+    if constexpr (oneapi::dpl::__internal::__can_avoid_placement_new_in_value_construct<_ValueType>)
+    {
+        oneapi::dpl::__internal::__ranges::__pattern_walk_n(
```

Comment:
Why do we even think that assignment is anyhow better than calling a value initialization over a memory? Why are we trying so hard to avoid placement new

Reply:
Let's continue the discussion in https://github.com/uxlfoundation/oneDPL/pull/2292#discussion_r2317827170.